_rules/text-contrast-widget.md

```diff
@@ -0,0 +1,95 @@
+---
+id: xxtodo
+name: Text imnside widget has minimum contrast
+rule_type: atomic
+description: |
+  TODO
+accessibility_requirements:
+  wcag20:1.4.3: # Contrast (Minimum)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+  wcag20:1.4.6: # Contrast (Enhanced)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - Accessibility Tree
+  - DOM Tree
+  - CSS Styling
+  - Language
+acknowledgments:
+  authors:
+    - Brian Bors
+    - Wilco Fiers
+---
+
+## Applicability
+
+The rule applies to any [visible][] character in a [text node][] that is a [child][] in the [flat tree][] of an HTML element, except if the [text node][] has an [ancestor][] in the [flat tree][] for which one of the following is true:
+
+- **disabled label**: the ancestor is used in the [accessible name][] of a `widget` that is [disabled][]; or
+- **disabled group**: the ancestor has a [semantic role][] of `group` and is [disabled][].
+
+and where the text node has at least one ancestor in the flat tree matching at least one of these [pseudo-classes](https://html.spec.whatwg.org/multipage/semantics-other.html#pseudo-classes): link, visited, enabled, checked, indeterminate, default, placeholder-shown, valid, invalid, in-range, out-of-range, required, optional, read-only, read-write.
```

Comment:
ARIA widgets don't get these pseudo classes. How would you propose handling them?

Reply:
Yes. Another Applicability could be "role inheriting from `widget`".
However, links can be `:link` or `:visited` and we want to check both (other elements have the same problem). Selecting by role isn't explicit about this.

A bit more research is needed on the possible combinations of these pseudo-classes (including none for a `<div role="link">`, I guess), to phrase it in a correct way. Care must be taken to not accidentally include elements that are `:disabled` (maybe an explicit exception is better).